yente/search/indexer.py

```diff
@@ -75,6 +87,89 @@ async def iter_entity_docs(
         except FollowTheMoneyException as exc:
             log.warning("Invalid entity: %s" % exc, data=data)
+    log.info(
+        "Indexed %d entities" % idx,
+        added=ops["ADD"],
+        modified=ops["MOD"],
+        deleted=ops["DEL"],
+    )
+
+
+async def rollover_index(
+    es: AsyncElasticsearch, alias: str, next_index: str, prefix: str
+) -> None:
+    """Remove all existing indices with a given prefix from the alias and
+    add the new one."""
+    actions = []
+    actions.append({"remove": {"index": f"{prefix}*", "alias": alias}})
+    actions.append({"add": {"index": next_index, "alias": alias}})
+    await es.indices.update_aliases(actions=actions)
+    log.info("Index is now aliased to: %s" % settings.ENTITY_INDEX, index=next_index)
+
+
+async def clone_index(es: AsyncElasticsearch, base_version: str, target_version: str):
+    """Create a copy of the index with the given name."""
+    if base_version == target_version:
+        raise ValueError("Cannot clone an index to itself.")
+    try:
+        await es.indices.put_settings(
+            index=base_version,
+            settings={"index.blocks.read_only": True},
+        )
+        await es.indices.delete(index=target_version, allow_no_indices=True)
+        await es.indices.clone(
+            index=base_version,
+            target=target_version,
+            body={
+                "settings": {"index": {"blocks": {"read_only": False}}},
+            },
+        )
+        log.info("Cloned index", base=base_version, target=target_version)
+    finally:
+        await es.indices.put_settings(
+            index=base_version,
+            settings={"index.blocks.read_only": False},
+        )
+
+
+async def create_index(es: AsyncElasticsearch, index: str):
+    """Create a new index with the given name."""
+    log.info("Create index", index=index)
+    try:
+        schemata = list(model.schemata.values())
+        mapping = make_entity_mapping(schemata)
+        await es.indices.create(
+            index=index,
+            mappings=mapping,
+            settings=INDEX_SETTINGS,
+        )
+    except BadRequestError as exc:
+        log.warning(
+            "Cannot create index: %s" % exc.message,
+            index=index,
+        )
+
+
+async def get_index_version(es: AsyncElasticsearch, dataset: Dataset) -> str | None:
+    """Return the currently indexed version of a given dataset."""
+    try:
+        resp = await es.indices.get_alias(name=settings.ENTITY_INDEX)
+    except NotFoundError:
+        return None
+    versions: List[str] = []
+    for index in resp.keys():
+        try:
+            ds, version = parse_index_name(index)
+            if ds == dataset.name:
+                versions.append(version)
+        except ValueError:
+            pass
+    if len(versions) == 0:
+        return None
+    # Return the oldest version of the index. If multiple versions are linked to the
```

Comment:
This is essentially everything in the Index class, right? If it takes the SearchProvider class instead then we'll save ourselves a headache when adding OpenSearch support.